### ISSUE TITLE
TryFunction Error in Base Application Event Subsriber to OnAfterSendEmail

### ISSUE THREAD

**Issue by pri-kise** [microsoft/ALAppExtensions#13375]
Please include the following with each issue:

**1. Describe the bug**
The BaseApp subscribes to events that are called inside a tryFunction

**2. To Reproduce**
The publisher Event inside the codeunit 8888 "Email Dispatcher"
```al
    [TryFunction]
    local procedure FireOnAfterSendEmail(MessageID: Guid; Success: Boolean)
    var
        Email: Codeunit Email;
    begin
        Email.OnAfterSendEmail(MessageID, Success);
    end;
```
The following event subscriber is implemented in the codeunit 1368 "Monitored Field Notification"
``` AL
    [EventSubscriber(ObjectType::Codeunit, Codeunit::Email, 'OnAfterSendEmail', '', false, false)]
    local procedure OnAfterSendEmail(MessageId: Guid; Status: Boolean)
    begin
        UpdateEmailStatus(MessageId, Status);
    end;

    procedure UpdateEmailStatus(MessageId: Guid; Status: Boolean)
    var
        ChangeLogEntry: Record "Change Log Entry";
    begin
        ChangeLogEntry.SetRange("Notification Message Id", MessageId);
        ChangeLogEntry.SetRange("Field Log Entry Feature", ChangeLogEntry."Field Log Entry Feature"::"Monitor Sensitive Fields");
        ChangeLogEntry.SetRange("Notification Status", ChangeLogEntry."Notification Status"::"Email Enqueued");
        if ChangeLogEntry.FindFirst() then begin
            if Status then
                ChangeLogEntry."Notification Status" := ChangeLogEntry."Notification Status"::"Email Sent"
            else
                ChangeLogEntry."Notification Status" := ChangeLogEntry."Notification Status"::"Sending Email Failed";

            ChangeLogEntry.Modify(); ///PROBLEM in Try Functions
        end;
    end;
```
**3. Expected behavior**
Could you please add a Background Task or something else asynchronusly, that will call this publisher event out of this try function. We need this in our applications that we can track the status of the email outbox messages.
And the BaseApp needs this, too.

**4. Actual behavior**
A clear and concise description of what happened accompanied by images, animations, or a link to a video showing the issue occurring

**5. Versions:**

- AL Language:
- Visual Studio Code:
- Business Central:
- List of Visual Studio Code extensions that you have installed:

### Final Checklist

Please remember to do the following:

* [x] Search the issue repository to ensure you are reporting a new issue

* [ ] Reproduce the issue after disabling all extensions except the AL Language extension

* [ ] Simplify your code around the issue to better isolate the problem

**Comment by pri-kise** [comment]
@AlexanderYakunin do I need to open a Support Request too, that this will be fixed for BC17 or will you provide a fix for this Error in "Monitored Field Notification" Subscriber?

This issue is related to the following issue #9553

**Comment by AlexanderYakunin** [comment]
@JesperSchulz - forwarding to your team

**Comment by GiorgosPa** [comment]
@pri-kise I am not sure yet whether we should change the event or the subscriber.
You can get unblocked by changing your subscriber to start a task for making the database changes.

**Comment by pri-kise** [comment]
@GiorgosPa okay this could maybe work for our sue case.
Do we need to open a support case that this will be in the Base Application?

**Comment by AlexanderYakunin** [comment]
You ask is not a hotfix, but design change. CSS and SE team are not doing design changes. This is a task for feature team working on Email module. And it should be done for Master branch first.

**Comment by GiorgosPa** [comment]
@pri-kise I think you will need a support case to get this hotfixed on 17.5.

**Comment by GiorgosPa** [comment]
Just noticed @AlexanderYakunin comment. Not sure what the process is let me find out and I will get back to you.

**Comment by GiorgosPa** [comment]
@pri-kise You need a support ticket to get any fix for BC17

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

**Comment by pri-kise** [comment]
@JesperSchulz 
As far as I can see.
This isn't fixed in the next major or minor in the BaseApp
```al
[EventSubscriber(ObjectType::Codeunit, Codeunit::Email, 'OnAfterSendEmail', '', false, false)]
local procedure OnAfterSendEmail(MessageId: Guid; Status: Boolean)
begin
    UpdateEmailStatus(MessageId, Status);
end;

procedure UpdateEmailStatus(MessageId: Guid; Status: Boolean)
var
    ChangeLogEntry: Record "Change Log Entry";
begin
    if IsNullGuid(MessageId) then
        exit;

    ChangeLogEntry.SetRange("Notification Message Id", MessageId);
    ChangeLogEntry.SetRange("Field Log Entry Feature", ChangeLogEntry."Field Log Entry Feature"::"Monitor Sensitive Fields");
    ChangeLogEntry.SetRange("Notification Status", ChangeLogEntry."Notification Status"::"Email Enqueued");
    if ChangeLogEntry.FindFirst() then begin
        if Status then
            ChangeLogEntry."Notification Status" := ChangeLogEntry."Notification Status"::"Email Sent"
        else
            ChangeLogEntry."Notification Status" := ChangeLogEntry."Notification Status"::"Sending Email Failed";

        ChangeLogEntry.Modify(); ///This is the problem
    end;
end;
```

**Comment by JesperSchulz** [comment]
Let me double-check! We'll get back to you.

**Comment by IhorHandziuk** [comment]
Hi @pri-kise, there isn't a change to the subscriber in BaseApp because the fix addressed the root cause: the `OnAfterSendEmail` is no longer raised from inside a `TryFunction` (please, see ["Email Dispatcher"](https://github.com/microsoft/ALAppExtensions/blob/main/Modules/System/Email/src/Email/EmailDispatcher.Codeunit.al) codeunit), so database write transactions will be allowed inside subscribers.

**Comment by pri-kise** [comment]
Hi @IhorHandziuk,
Thanks for the clarification. Seems, that I missed that change.
